src/graph/creation.rs

```diff
@@ -0,0 +1,254 @@
+// License: MIT
+// Copyright © 2024 Frequenz Energy-as-a-Service GmbH
+
+//! Methods for creating [`ComponentGraph`] instances from given components and
+//! connections.
+
+use petgraph::graph::DiGraph;
+
+use crate::{component_category::CategoryPredicates, Edge, Error, Node};
+
+use super::{validation::validate, ComponentGraph, EdgeMap, NodeIndexMap};
+
+/// `ComponentGraph` instantiation.
+impl<N, E> ComponentGraph<N, E>
+where
+    N: Node,
+    E: Edge,
+{
+    /// Creates a new [`ComponentGraph`] from the given components and connections.
+    ///
+    /// Returns an error if the graph is invalid.
+    pub fn try_new<NodeIterator: IntoIterator<Item = N>, EdgeIterator: IntoIterator<Item = E>>(
+        components: NodeIterator,
+        connections: EdgeIterator,
+    ) -> Result<Self, Error> {
+        let (graph, indices) = Self::create_graph(components)?;
+        let root_id = Self::find_root(&graph)?.component_id();
+
+        let mut cg = Self {
+            graph,
+            node_indices: indices,
+            root_id,
+            edges: EdgeMap::new(),
+        };
+        cg.add_connections(connections)?;
+
+        validate(&cg)?;
+
+        Ok(cg)
+    }
+
+    fn find_root(graph: &DiGraph<N, ()>) -> Result<&N, Error> {
+        let mut roots_iter = graph.raw_nodes().iter().filter(|n| n.weight.is_grid());
+
+        let root = roots_iter
+            .next()
+            .map(|n| &n.weight)
+            .ok_or_else(|| Error::invalid_graph("No grid component found."))?;
+
+        if roots_iter.next().is_some() {
+            return Err(Error::invalid_graph("Multiple grid components found."));
+        }
+
+        Ok(root)
+    }
+
+    fn create_graph(
+        components: impl IntoIterator<Item = N>,
+    ) -> Result<(DiGraph<N, ()>, NodeIndexMap), Error> {
+        let mut graph = DiGraph::new();
+        let mut indices = NodeIndexMap::new();
+
+        for component in components {
+            let cid = component.component_id();
+
+            if component.is_unspecified() {
+                return Err(Error::invalid_component(format!(
+                    "ComponentCategory not specified for component: {cid}"
+                )));
```

Comment:
I wonder if having a unspecified component (or inverter) is an error fatal enough to completely refuse to create the graph. Couldn't we just emit an error and continue and somehow deal with unspecified components later, depending on the operation we want to do? Specially if a unspecified component is a leaf it looks like it might be harmless to ignore it (maybe we'll get wrong consumption or production, but better than nothing, right?)

Reply:
I agree, will make it configurable in a subsequent PR.